tfhe/src/keycache/mod.rs

```diff
@@ -214,4 +214,73 @@ pub mod utils {
             }
         }
     }
+
+    impl<P, K, S> KeyCache<P, K, S>
+    where
+        P: Copy + PartialEq + NamedParam,
+        S: PersistentStorage<P, K>,
+    {
+        pub fn get_with_closure<C: FnMut(P) -> K>(&self, param: P, c: &mut C) -> SharedKey<K> {
+            self.with_closure(param, c)
+        }
+
+        pub fn with_closure<C>(&self, param: P, c: &mut C) -> SharedKey<K>
+        // pub fn with_closure<C, R: Clone>(&self, param: P, c: C) -> R
+        where
+            C: FnMut(P) -> K,
+        {
+            let load_from_persistent_storage = || {
+                // we check if we can load the key from persistent storage
+                let persistent_storage = &self.persistent_storage;
+                let maybe_key = persistent_storage.load(param);
+                maybe_key.map_or_else(
+                    || {
+                        let key = c(param);
+                        persistent_storage.store(param, &key);
+                        key
+                    },
+                    |key| key,
+                )
+            };
+
+            let try_load_from_memory_and_init = || -> Result<_, ()> {
+                // we only hold a read lock for a short duration to find the key
+                let maybe_shared_cell = {
+                    let memory_storage = self.memory_storage.read().unwrap();
+                    memory_storage
+                        .iter()
+                        .find(|(p, _)| *p == param)
+                        .map(|param_key| param_key.1.clone())
+                };
+
+                maybe_shared_cell.map_or_else(
+                    || {
+                        {
+                            // we only hold a write lock for a short duration to push the lazily
+                            // evaluated key without actually evaluating the key
+                            let mut memory_storage = self.memory_storage.write().unwrap();
+                            if !memory_storage.iter().any(|(p, _)| *p == param) {
+                                let shared_key = SharedKey {
+                                    inner: Arc::new(OnceLock::new()),
+                                };
+                                shared_key.inner.get_or_init(load_from_persistent_storage);
+                                memory_storage.push((param, shared_key));
+                            }
+                        }
+                        let memory_storage = self.memory_storage.read().unwrap();
+                        let shared_cell = memory_storage
+                            .iter()
+                            .find(|(p, _)| *p == param)
+                            .map(|param_key| param_key.1.clone())
+                            .unwrap();
+
+                        Ok(shared_cell)
+                    },
+                    Ok,
+                )
+            };
+
+            try_load_from_memory_and_init().ok().unwrap()
+        }
+    }
```

Comment:
Now that we have this, shouldn't we able to implement `Keycache::get` as 
```rust
self.get_with_closure(params, K::from)
```
Or something similar

Reply:
sounds very likely indeed